### DISCUSSION TITLE
TypeScript - Movies - Interface, Objects, For Loop

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/angular#146] [Tasks]
# Task Description
You are tasked with creating a program that stores information about a list of movies in an array of objects. Your program should allow the user to add, access, and delete information about the movies.

# Requirements
- Create an interface called Movie with the following properties:
    - `title` of type `string`
    - `director` of type `string`
    - `releaseYear` of type `number`
    - `rating` of type `number`
- Create an empty array called movieList that will store all the movies.
- Use the Movie interface to create an object for each movie in the movieList array. The movies in the array should be:
    - Title: The Matrix, Director: The Wachowskis, Release Year: 1999, Rating: 9
    - Title: Inception, Director: Christopher Nolan, Release Year: 2010, Rating: 8
    - Title: The Shawshank Redemption, Director: Frank Darabont, Release Year: 1994, Rating: 10

- Use a for loop to iterate through the movieList array and output the title and rating of each movie using the console.log() function.
- Create a function called addMovie() that accepts a Movie object as a parameter and adds it to the movieList array.
- Create a function called deleteMovie() that accepts a string parameter representing the title of the movie to delete. The function should remove the movie from the movieList array.
- Create a function called getMovieByDirector() that accepts a string parameter representing the name of the director. The function should iterate through the movieList array and return an array of all the movies with the specified director.
- You should use TypeScript to define the types of the input and output parameters for the functions.

# Note

- We create three functions: `addMovie()`, `deleteMovie()`, and `getMovieByDirector()`. The `addMovie()` function accepts a `Movie` object as a parameter and adds it to the `movieList` array. The `deleteMovie()` function accepts a `string` parameter representing the title of the movie to delete and removes it from the `movieList` array. The `getMovieByDirector()` function accepts a `string` parameter representing the name of the director and returns an array of all the movies with the specified director.
- We also use TypeScript to define the types of the input and output parameters for the functions. This helps ensure that the functions are used correctly and avoid common programming errors.

**Comment by saugatacr7** [comment]
interface Movie{
    name: String,
    Director: String,
    Release Year: number,
    Rating: number,
}
let movieList=[];
let movieList=new movieList();
window.movieList = window.movieList || [];
console.log()

let movieList = [
  {
    "Title": "Matrix",
    "Director": "The Wachowskis",
    "Release Year": 1999,
    "Rating": 9
  },
  {
    "Title": "Inception",
    "Director": "Christopher Nolan",
    "Release Year": 2010,
    "Rating": 8
  }
  {
      "Title": "The Shawshank Redemption",
    "Director": "Frank Darabont",
    "Release Year": 1994,
    "Rating": 10
  }
  
  for(let key in movieList){
      console.log('key' ':' 'movieList[Title]');
      console.log('key' ':' 'movieList[Rating]');
      
      
  }

  function pushFunction() {   
      //add new element in array        
      movieList.push(" Avengers", " Prison Break", " The Spy"); 
      console.log('key' ':' 'movieList[]);
      
  }
      function popFunction(){
          movieList.pop("Title")
          console.log(console.log('key' ':' 'movieList[]);)
      }

**Comment by bipin-coded** [comment]
## Code

```typescript
interface Movie {
  title: string;
  director: string;
  releaseYear: number;
  rating: number;
}

let movieList: Movie[] = [
  {
    title: "The Matrix",
    director: "The Wachowskis",
    releaseYear: 1999,
    rating: 9,
  },
  {
    title: "Inception",
    director: "Christopher Nolan",
    releaseYear: 2010,
    rating: 8,
  },
  {
    title: "The Shawshank Redemption",
    director: "Frank Darabont",
    releaseYear: 1994,
    rating: 10,
  }
];

for (let movie in movieList) {
  console.log(movieList[movie].title, ':', movieList[movie].rating);
}


let addMovie = function(movie: Movie) {
  movieList.push(movie);
  return movieList;
}

let deleteMovie = function(title: string) {
  return movieList.filter(movie => {
    return title !== movie['title']
  }); 
}

let getMovieByDirector =  function(director: string) {
  return movieList.filter(movie => {
    return director === movie['director']
  });
}
```

## Output
![Screenshot (4)](https://user-images.githubusercontent.com/126976134/232326822-11446a0a-11de-419a-9695-f1d9df663259.png)

**Comment by chukku10** [comment]
interface Movie{
name: String,
Director: String,
Release Year: number,
Rating: number,
}
let movieList=[];
let movieList=new movieList();
window.movieList = window.movieList || [];
console.log()

let movieList = [
{
"Title": "Matrix",
"Director": "The Wachowskis",
"Release Year": 1999,
"Rating": 9
},
{
"Title": "Inception",
"Director": "Christopher Nolan",
"Release Year": 2010,
"Rating": 8
}
{
"Title": "The Shawshank Redemption",
"Director": "Frank Darabont",
"Release Year": 1994,
"Rating": 10
}

for(let key in movieList){
console.log('key' ':' 'movieList[Title]');
console.log('key' ':' 'movieList[Rating]');

}

function pushFunction() {
//add new element in array
movieList.push(" Avengers", " Prison Break", " The Spy");
console.log('key' ':' 'movieList[]);

}
function popFunction(){
movieList.pop("Title")
console.log(console.log('key' ':' 'movieList[]);)
}

**Comment by SathyaGanes** [comment]
Code

```
interface Movie {
  title: string;
  director: string;
  releaseYear: number;
  rating: number;
}

const movieList: Movie[] = [
  {
    title: "The Matrix",
    director: "The Wachowskis",
    releaseYear: 1999,
    rating: 9,
  },
  {
    title: "Inception",
    director: "Christopher Nolan",
    releaseYear: 2010,
    rating: 8,
  },
  {
    title: "The Shawshank Redemption",
    director: "Frank Darabont",
    releaseYear: 1994,
    rating: 10,
  },
];


for (const movie of movieList) {
  console.log(`Title: ${movie.title}, Rating: ${movie.rating}`);
}

function addMovie(movie: Movie): void {
  movieList.push(movie);
}

function deleteMovie(title: string): void {
  const index = movieList.findIndex((movie) => movie.title === title);
  if (index !== -1) {
    movieList.splice(index, 1);
  }
}

function getMovieByDirector(director: string): Movie[] {
  return movieList.filter((movie) => movie.director === director);
}


const newMovie: Movie = {
  title: "The Dark Knight",
  director: "Christopher Nolan",
  releaseYear: 2008,
  rating: 9,
};
addMovie(newMovie);
console.log(movieList); // Output: updated movieList array with new movie

deleteMovie("Inception");
console.log(movieList); // Output: updated movieList array without Inception

const moviesByNolan = getMovieByDirector("Christopher Nolan");
console.log(moviesByNolan); // Output: array of movies directed by Christopher Nolan



Output

Title: The Matrix, Rating: 9
Title: Inception, Rating: 8
Title: The Shawshank Redemption, Rating: 10
[
  {
    title: 'The Matrix',
    director: 'The Wachowskis',
    releaseYear: 1999,
    rating: 9
  },
  {
    title: 'Inception',
    director: 'Christopher Nolan',
    releaseYear: 2010,
    rating: 8
  },
  {
    title: 'The Shawshank Redemption',
    director: 'Frank Darabont',
    releaseYear: 1994,
    rating: 10
  },
  {
    title: 'The Dark Knight',
    director: 'Christopher Nolan',
    releaseYear: 2008,
    rating: 9
  }
]
[
  {
    title: 'The Matrix',
    director: 'The Wachowskis',
    releaseYear: 1999,
    rating: 9
  },
  {
    title: 'The Shawshank Redemption',
    director: 'Frank Darabont',
    releaseYear: 1994,
    rating: 10
  },
  {
    title: 'The Dark Knight',
    director: 'Christopher Nolan',
    releaseYear: 2008,
    rating: 9
  }
]
[
  {
    title: 'Inception',
    director: 'Christopher Nolan',
    releaseYear: 2010,
    rating: 8
  },
  {
    title: 'The Dark Knight',
    director: 'Christopher Nolan',
    releaseYear: 2008,
    rating: 9
  }
]

```

**Comment by starvikas1988** [comment]
interface Movie_vikas {
    title: string;
    director: string;
    releaseYear: number;
    rating : number;
  }

  let movieList: Movie_vikas[] = [
    {
      title: "The Matrix",
      director: "The Wachowskis",
      releaseYear: 1999,
      rating: 9,
    },
    {
      title: "Inception",
      director: "Christopher Nolan",
      releaseYear: 2010,
      rating: 8,
    },
    {
      title: "The Shawshank Redemption",
      director: "Frank Darabont",
      releaseYear: 1994,
      rating: 10,
    },
  ]; 

  for(let key in movieList){
    //console.log(movieList[key].title," ",movieList[key].director," ",movieList[key].releaseYear," ",movieList[key].rating);
  }

  function addMovie(x: Movie_vikas): any {
    x.title = this.title;
    x.director = this.year;
    x.releaseYear = this.year;
    x.rating = this.year;
    movieList.push(x);

  }

  addMovie({
      title: "The Gambller returns",
      director: "Vikas Pandey",
      releaseYear: 2023,
      rating: 10,
  });

  for(let key in movieList){
    console.log(movieList[key].title," ",movieList[key].director," ",movieList[key].releaseYear," ",movieList[key].rating);
  }

**Comment by AratiKumari123** [comment]
```typescript
interface Movie {
  title: string;
  director: string;
  releaseYear: number;
  rating: number;
}

const movieList: Movie[] = [
  {
    title: "The Matrix",
    director: "The Wachowskis",
    releaseYear: 1999,
    rating: 9,
  },
  {
    title: "Inception",
    director: "Christopher Nolan",
    releaseYear: 2010,
    rating: 8,
  },
  {
    title: "The Shawshank Redemption",
    director: "Frank Darabont",
    releaseYear: 1994,
    rating: 10,
  },
];


for (const movie of movieList) {
  console.log(`Title: ${movie.title}, Rating: ${movie.rating}`);
}

function addMovie(movie: Movie): void {
  movieList.push(movie);
}

function deleteMovie(title: string): void {
  const index = movieList.((movie) => movie.title === title);
  if (index !== -1) {
    movieList.splice(index, 1);
  }
}

function getMovieByDirector(director: string): Movie[] {
  return movieList.filter((movie) => movie.director === director);
}


const newMovie: Movie = {
  title: "The Dark Knight",
  director: "Christopher Nolan",
  releaseYear: 2008,
  rating: 9,
};
addMovie(newMovie);
console.log(movieList); // Output: updated movieList array with new movie

deleteMovie("Inception");
console.log(movieList); // Output: updated movieList array without Inception

const moviesByNolan = getMovieByDirector("Christopher Nolan");
console.log(moviesByNolan); // Output: array of movies directed by Christopher Nolan

```

**Comment by basu-cse** [comment]
interface Movie {
  title: string;
  director: string;
  releaseYear: number;
  rating: number;
}

const movieList: Movie[] = [
  {
    title: "The Matrix",
    director: "The Wachowskis",
    releaseYear: 1999,
    rating: 9,
  },
  {
    title: "Inception",
    director: "Christopher Nolan",
    releaseYear: 2010,
    rating: 8,
  },
  {
    title: "The Shawshank Redemption",
    director: "Frank Darabont",
    releaseYear: 1994,
    rating: 10,
  },
];

for (const movie of movieList) {
  console.log(`Title: ${movie.title}, Rating: ${movie.rating}`);
}

function addMovie(movie: Movie): void {
  movieList.push(movie);
}

function deleteMovie(title: string): void {
  const index = movieList.findIndex((movie) => movie.title === title);
  if (index !== -1) {
    movieList.splice(index, 1);
  }
}

function getMovieByDirector(director: string): Movie[] {
  return movieList.filter((movie) => movie.director === director);
}

**Comment by RevathiSundaramoorthy** [comment]
## Code
```typescript
interface Movie {
    title: string,
    director: string,
    releaseYear: number,
    rating: number
}

let movieList: Movie[] = [];
movieList.push({title: 'The Matrix', director: 'The Wachwskis', releaseYear: 1999, rating:9});
//add movie using function addMovie()
addMovie({title: 'Inception', director: 'Christopher Nolan', releaseYear: 2010, rating:8});
addMovie({title: 'The Shawshank Redemption', director: 'Frank Darabont', releaseYear: 1994, rating:10});
addMovie({title: 'The Prestige', director: 'Christopher Nolan', releaseYear: 2006, rating:7});
addMovie({title: 'Following', director: 'Christopher Nolan', releaseYear: 1998, rating:7.1});
//display title and rating
for (let movie in movieList) {
    console.log("Title : " + movieList[movie].title)
    console.log("Rating : " + movieList[movie].rating);
}
//delete movie using function deleteMovie()
deleteMovie('The Prestige');
//get movie list be director
let movieListByDirector = getMovieByDirector('Christopher Nolan');
console.log("Movies directed by Christopher Nolan: " + movieListByDirector);
//function to add movie
function addMovie(movie: Movie) {
    movieList.push(movie);
    return;
}
function deleteMovie(title: string) {
    for (let movie in movieList) {
        if (movieList[movie].title == title) {
            delete movieList[movie];
            console.log ("Deleted the movie: " + title);
        }
    }
    return;
}
function getMovieByDirector(director: string): string[] {
    let moviesListDirector: string[] = []
    for (let movie in movieList) {
        if (movieList[movie].director == director) {
            moviesListDirector.push(movieList[movie].title);
        }
    }
    return moviesListDirector;
}
```
## Screenshot
![image](https://user-images.githubusercontent.com/126975777/232335484-865479d3-40ec-40cb-8fa7-567833075090.png)

**Comment by vk0027** [comment]
interface Movie {
  title: string;
  director: string;
  releaseYear: number;
  rating: number;
}

const movieList: Movie[] = [
  {
    title: 'The Matrix',
    director: 'The Wachowskis',
    releaseYear: 1999,
    rating: 9,
  },
  {
    title: 'Inception',
    director: 'Christopher Nolan',
    releaseYear: 2010,
    rating: 8,
  },
  {
    title: 'The Shawshank Redemption',
    director: 'Frank Darabont',
    releaseYear: 1994,
    rating: 10,
  },
];

function addMovie(movie: Movie): void {
  movieList.push(movie);
}

function deleteMovie(title: string): void {
  const index = movieList.findIndex((movie) => movie.title === title);
  if (index !== -1) {
    movieList.splice(index, 1);
  }
}

function getMovieByDirector(director: string): Movie[] {
  return movieList.filter((movie) => movie.director === director);
}

for (const movie of movieList) {
  console.log(`Title: ${movie.title}, Rating: ${movie.rating}`);
}

**Comment by mownica-p** [comment]
# Code
```TypeScript
interface Movie {
    title: string;
    director: string;
    releaseYear: number;
    rating: number;
  }
  
  const movieList: Movie[] = [
    {
      title: "The Matrix",
      director: "The Wachowskis",
      releaseYear: 1999,
      rating: 9,
    },
    {
      title: "Inception",
      director: "Christopher Nolan",
      releaseYear: 2010,
      rating: 8,
    },
    {
      title: "The Shawshank Redemption",
      director: "Frank Darabont",
      releaseYear: 1994,
      rating: 10,
    },
  ];
  
  
  for (const movie of movieList) {
    console.log(`Title: ${movie.title}, Rating: ${movie.rating}`);
  }

//Add New Movie  
  function addMovie(movie: Movie): void {
    movieList.push(movie);
  }
  const newMovie: Movie = {
    title: "The Dark Knight",
    director: "Christopher Nolan",
    releaseYear: 2008,
    rating: 9,
  };
  addMovie(newMovie);
  console.log(movieList);

//Delete Movie    
  function deleteMovie(title: string): void {
    let index = -1;
    for (let i = 0; i < movieList.length; i++) {
      if (movieList[i].title === title) {
        index = i;
        break;
      }
    }
    if (index !== -1) {
      movieList.splice(index, 1);
    }
  }

deleteMovie("Inception");
console.log(movieList); 

//Get Movie by Directory    
  function getMovieByDirector(director: string): Movie[] {
    return movieList.filter((movie) => movie.director === director);
  }

  const moviesByNolan = getMovieByDirector("The Wachowskis");
  console.log(moviesByNolan); 
```
# ScreenShot
![image](https://user-images.githubusercontent.com/126975660/233148620-234f6866-9202-4baa-9c0b-03b990bb0b05.png)
![image](https://user-images.githubusercontent.com/126975660/233148933-0b103cf7-dff9-4a3b-b8bb-3b88c0b87453.png)

**Comment by Prathmesh06** [comment]
# Code
```js
interface movie {
    Title : String;
    director : String;
    releaseYear: number;
    rating : number;
}

let movieList : movie[] = [
    {
        Title: "The Matrix",
        director: "The Wachowskis",
        releaseYear: 1998,
        rating: 9,
    },
     {
        Title: "Inception",
        director: "Christopher Nolan",
        releaseYear: 2010,
        rating: 8,
     },
     {
        Title: "The Shawshank Redemption",
        director: "Frank Darabont",
        releaseYear: 1994,
        rating: 10,
     },
];

console.log(movieList);

for (let key in movieList) {
    console.log(`title : ${movieList[key].Title}, rating : ${movieList[key].rating}`);
    console.log("Title : " + movieList[key].Title + ", Rating : " + movieList[key].rating);
}

const newMovie : movie = {
    Title: "Lost in Space",
    director: "Christan Lopise",
    releaseYear: 2021,
    rating: 9,
}

function addMovie(Movie : movie) {
    movieList.push(Movie);
    console.log("New movie added at list : " , movieList);
}

console.log(newMovie);

addMovie(newMovie);

function deleteMovie(movieName : string) {
    for (let movie in movieList) {
        if (movieName == movieList[movie].Title) {
            delete movieList[movie];
            console.log("This",movieName, "deleted  from database : ", movieList);
        } else {
            console.log(`Movie not found: ${movieName}`);
        }
    }
}

deleteMovie("Inception");

function getMovieByDirector(directorName : string) {
    for(let director in movieList) {
        if(movieList[director].director == directorName) {
            console.log("This director is directed" , directorName, movieList[director] );
        } else {
            console.log("the director is not found");
        }
    }
}

getMovieByDirector("The Wachowskis");
```

# output
![image](https://user-images.githubusercontent.com/58461727/233826967-05b600e8-4ebf-4a8d-8b1b-7094c45e29f1.png)
![image](https://user-images.githubusercontent.com/58461727/233827002-3751064b-59fa-483c-82ac-b8470496192b.png)
![image](https://user-images.githubusercontent.com/58461727/233827028-e2f9a446-1975-433b-8be6-5e8c2869ae84.png)

**Comment by sam2elect** [comment]
# CODE
```ts
interface Movie {
    title: string;
    director: string;
    releaseYear: number;
    rating: number;
  }
  
  const movieList: Movie[] = [
    { title: "The Matrix", director: "The Wachowskis", releaseYear: 1999, rating: 9 },
    { title: "Inception", director: "Christopher Nolan", releaseYear: 2010, rating: 8 },
    { title: "The Shawshank Redemption", director: "Frank Darabont", releaseYear: 1994, rating: 10 }
  ];
  
  // Output title and rating of each movie in movieList
  for (const movie of movieList) {
    console.log(`Title: ${movie.title}, Rating: ${movie.rating}`);
  }
  
  function addMovie(movie: Movie): void {
    movieList.push(movie);
  }
  
  function deleteMovie(title: string): void {
    const index = movieList.findIndex(movie => movie.title === title);
    if (index !== -1) {
      movieList.splice(index, 1);
    }
  }
  
  function getMovieByDirector(director: string): Movie[] {
    return movieList.filter(movie => movie.director === director);
  }
  
```
# SCREENSHORT

![image](https://user-images.githubusercontent.com/85131631/235346419-671c8114-b967-409e-87c7-4d6594fa21ef.png)

**Comment by kumarneeraj89** [comment]
# Code

```typescript
interface Movie{
    title: string;
    director: string;
    releaseYear: number;
    rating: number;
}

let movieList = [] = [{
    title: 'The Matrix',
    director: 'The Wachowskis', 
    releaseYear: 1999, 
    rating: 9
},
   { title: 'Inception',
    director: 'Christopher Nolan', 
    releaseYear: 2010, 
    rating: 8
   },
   {
    title: 'The Shawshank Redemption',
    director: 'Frank Darabont', 
    releaseYear: 1994, 
    rating: 10
   },
   {
   title: 'Bound',
   director: 'The Wachowskis', 
   releaseYear: 1996, 
   rating: 7.3
   }
]

for(let i = 0; i < movieList.length; i++){
    console.log(`the title and rating of each movie are ${movieList[i].title} & ${movieList[i].rating}`);
}

function addMovie(movie: Movie) {
    movieList.push(movie);
    return true;
}

function deleteMovie(movietitle: string) {

    for(let i = 0; i < movieList.length; i++) {
        if(movietitle == movieList[i].title){
                movieList.splice(i, 1);
                }
        }
   // const index = movieList.findIndex((movie) => movie.title === movietitle);
   //if (index !== -1) {
//movieList.splice(index, 1);
    }

    const directorlist: string[] = [];
    function getMovieByDirector(movieDirectorName: string) {
        

        for(let i = 0; i < movieList.length; i++) {
           
            if(movieDirectorName == movieList[i].director){
                directorlist.push(movieList[i].title);
                    }
            }
       // const index = movieList.findIndex((movie) => movie.title === movietitle);
       //if (index !== -1) {
    //movieList.splice(index, 1);
        }



let newMovie = {  title: "The Dark Knight",
director: "Christopher Nolan",
releaseYear: 2008,
rating: 9,}

addMovie(newMovie);

console.log('-------------------------------- List of movies after addition');
console.log(movieList);

deleteMovie("Inception")
console.log('-------------------------------- List of movies after deletion');
console.log(movieList);

console.log('--------------------------------')
getMovieByDirector("The Wachowskis");
console.log(directorlist);
```

# Output

```
the rating of movie name The Matrix is 9
the rating of movie name Inception is 8
the rating of movie name The Shawshank Redemption is 10
the rating of movie name Bound is 7.3
-------------------------------- List of movies after addition
[
  {
    title: 'The Matrix',
    director: 'The Wachowskis',
    releaseYear: 1999,
    rating: 9
  },
  {
    title: 'Inception',
    director: 'Christopher Nolan',
    releaseYear: 2010,
    rating: 8
  },
  {
    title: 'The Shawshank Redemption',
    director: 'Frank Darabont',
    releaseYear: 1994,
    rating: 10
  },
  {
    title: 'Bound',
    director: 'The Wachowskis',
    releaseYear: 1996,
    rating: 7.3
  },
  {
    title: 'The Dark Knight',
    director: 'Christopher Nolan',
    releaseYear: 2008,
    rating: 9
  }
]
-------------------------------- List of movies after deletion
[
  {
    title: 'The Matrix',
    director: 'The Wachowskis',
    releaseYear: 1999,
    rating: 9
  },
  {
    title: 'The Shawshank Redemption',
    director: 'Frank Darabont',
    releaseYear: 1994,
    rating: 10
  },
  {
    title: 'Bound',
    director: 'The Wachowskis',
    releaseYear: 1996,
    rating: 7.3
  },
  {
    title: 'The Dark Knight',
    director: 'Christopher Nolan',
    releaseYear: 2008,
    rating: 9
  }
]
--------------------------------
[ 'The Matrix', 'Bound' ]

[Done] exited with code=0 in 0.618 seconds
```
